### PR TITLE
[RLlib] PPO, APPO, and DD-PPO code cleanup.

### PR DESCRIPTION
**Pull request by sven1977**
Cleanup of PPO, APPO, and DD-PPO code (equivalent to previous PG one):
- add type annotations.
- get rid of loss class.
- add comments and explanations to everything we do in PPO.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Review by ericl — APPROVED** [review]
This looks great. @michaelzhiluo @richardliaw any further suggestions?

**Review by richardliaw — COMMENTED** [review]
(some comments)

File: rllib/policy/tf_policy_template.py [hunk 1]
```diff
@@ -10,60 +10,70 @@
...
-        mixins: Optional[List[type]] = None,
-        get_batch_divisibility_req: Optional[Callable[[Policy], int]] = None,
-        obs_include_prev_action_reward: bool = True):
+    name: str,
+    *,
+    loss_fn: Callable[[
+                          Policy, ModelV2, Type[TFActionDistribution],
+                          SampleBatch
+                      ], Union[TensorType, List[TensorType]]],
+    get_default_config: Optional[Callable[[None],
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/policy/tf_policy_template.py]: :D

File: rllib/agents/ppo/ddppo.py [hunk 2]
```diff
@@ -67,36 +75,64 @@
...
+
 # __sphinx_doc_end__
 # yapf: enable
 
 
 def validate_config(config):
+    """Validates the Trainer's config dict.
+
+    Args:
+        config (TrainerConfigDict): The Trainer's config to check.
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/ppo/ddppo.py:87]: It's just a `Dict`, really, but with the assumption that it's an actual Trainer config (not just an arbitrary dict).

File: rllib/agents/ppo/ppo.py [hunk 3]
```diff
@@ -159,23 +174,76 @@ def update(pi, pi_id):
...
+            "{} iterations for your value ".format(rew_scale) +
+            "function to converge. If this is not intended, consider "
+            "increasing `vf_clip_param`.")
+
+    return result
+
+
+def execution_plan(workers: WorkerSet,
+                   config: TrainerConfigDict) -> LocalIterator[dict]:
+    """Execution plan of the PPO algorithm. Defines the distributed dataflow.
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/ppo/ppo.py:228]: done

File: rllib/agents/ppo/ppo.py [hunk 4]
```diff
@@ -159,23 +174,76 @@ def update(pi, pi_id):
...
+
+    Args:
+        workers (WorkerSet): The WorkerSet for training the Polic(y/ies)
+            of the Trainer.
+        config (TrainerConfigDict): The trainer's configuration dict.
+
+    Returns:
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/ppo/ppo.py:235]: ok.

**Pull request merged**